Add configurable throw exception when request body is larger than request cache size

serii833 opened the pull request:
My intent for this PR is to add exception throwing when request body is larger than request cache.
This feature is disabled by default. 
When enableRequestTooLongException() javalin will throw exception if request cache is too small.
to my mind to get exception make it more predictable then to get empty body.

Playacem commented:
I personally dislike the inconsistency in naming. You call it `RequestTooLongException`, `exceptionOnRequestCacheIsSmall`.

Playacem commented on src/main/java/io/javalin/core/CachedRequestWrapper.kt:
```diff
@@ -23,12 +23,16 @@ class CachedRequestWrapper(request: HttpServletRequest, private val maxCacheSize
-            }
+    override fun getInputStream(): ServletInputStream {
+        if (exceptionOnRequestCacheIsSmall && maxCacheSize < size)
+            throw Exception("Request cache size is too small for request body.")
```
Throwing Exception directly feels wrong for me personally. Is there a better Exception available?

serii833 replied:
fixed in 1fa7d2106cf60572c86edcef229c77e667e561df

Playacem commented on src/test/java/io/javalin/TestRequestTooLong.kt:
```diff
@@ -0,0 +1,43 @@
+            }
+
+            val longPostBody = "{\"some_long_string\":\"loooooongteext\"}";
+            assertThat(http.post("/long-post-body").body(longPostBody).asString().body, `is`("java.lang.Exception: Request cache size is too small for request body.") )
```
checking for the exact exception message feels brittle and not very clean too me.

serii833 replied:
fixed in 1fa7d2106cf60572c86edcef229c77e667e561df

Playacem commented on src/main/java/io/javalin/Javalin.java:
```diff
@@ -165,7 +166,8 @@ public JavalinServlet createServlet() {
             singlePageHandler,
-            new JettyResourceHandler(staticFileConfig, jettyServer, ignoreTrailingSlashes)
+            new JettyResourceHandler(staticFileConfig, jettyServer, ignoreTrailingSlashes),
+            exceptionOnRequestCacheIsSmall
```
Did you check the EmbeddedJavalin class and if it requires any changes?

serii833 replied:
initially I wanted to make this parameter (to enable exception) unobtrusive. but I guess it's better to make it as normal parameter of javalin. 
I added it to createServlet of EmbeddedJavalin

Playacem commented on src/main/java/io/javalin/Javalin.java:
```diff
@@ -427,6 +429,17 @@ public Javalin maxBodySizeForRequestCache(long bodySizeInBytes) {
+     * The default value is false.
+     * The method must be called before {@link Javalin#start()}.
+     */
+    public Javalin enableRequestTooLongException() {
```
Do we need to override this method in EmbeddedJavalin?

serii833 replied:
no, there is no need to override this method

tipsy commented:
Thanks @serii833, I appreciate the effort you've put into this, but how useful is this feature when it's disabled by default?

serii833 commented:
@tipsy I can make it enabled by default if you agree with such behavior of framework. For me personally would be better to get an exception instead of empty body. It's more clear what is happening.

tipsy commented:
@serii833 I think your proposal makes sense as default behavior, but it would break backwards compatibility. I propose you rewrite the PR to throw exceptions by default (I don't think it needs to be possible to disable the behavior), but the merge will have to wait until Javalin 3.0.

Playacem commented:
We could add this feature now with the default being disabled and change the default to enabled with the release of Javalin 3.0. That way those who want this feature earlier could opt-in as soon as we release the change.

tipsy commented:
We could, but would you seek out and enable this feature without knowing about it? The current behavior is the default behavior for Servlets. 

What if we just log that the request was too big for the cache?

Playacem commented:
Maybe log it and advertise this option in the log message?

tipsy commented:
I'm trying to figure out how this is supposed to work. What are you supposed to do with these exceptions?

tipsy commented:
But both of this two options require you to deploy a new version of the app, so are these really runtime exceptions? Wouldn't it be enough to `log.warn` of the

serii833 commented:
sorry, I'm a little bit lost. You're talking about ways to implement this in javalin2 or about having such exception in general?

tipsy commented:
I'm not talking about implementation, but usage: How do you handle these exceptions? What you describe (increase cache size or go stream mode) requires you to make code changes, so what's the point of throwing a runtime exception?

tipsy commented:
> @serii833 body() is not good anyways, cause empty body is not what I expect to get.
therefore I'll get an exception somewhere down in my code or I can get some weird behavior of my code.

Okay, let's say the Javalin now throws `CacheTooSmallException`, can you show me how you would handle this exception?

serii833 commented:
in simplified case I'll handle it like this
```java
app.exception(Exception::class.java) { e, ctx ->
        logger.error(e) { "" }
        ctx.status(500)
}
```
then checking log manually or with some fancy stuff like elasticsearch with alerting I will see that I've got CacheTooSmallException. 
And this is the most important that I will see CacheTooSmallException. 
not some other exception (jackson, my code...) that will not give a clue that I need to increase cache.
this exception should guide me to docs saying that I need to increase cache size or go with "stream" mode. 


My real case was
everything was ok in development. 
we didn't use large requests and could consume body several times (first for bodyAsClass, second to log entire body for debugging). 
when we switched to preproduction we were hit with some large requests that we needed to debug from application logic perspective.  but it turns out that log did not contain any errors and logged body was empty.
When I changed code to first log body and than do bodyAsClass I began to get "could not parse" from jackson.
It was some hours of frustration before I opened javalin sources and figured it out. :)

I don't think that ```log.warn``` will be enough. ```log.warn``` is often not respected by developers and admins as much as exception.

serii833 commented:
Accidentally closed the PR

tipsy commented:
I agree with you that the current behavior is bad, but isn't throwing an exception is less helpful than logging a warning/error? In your exception handler you're just logging and setting status to 500, maybe that could be the default behavior? What else could you want to do for this exception?

tipsy commented:
@serii833 I still don't think the exception is a good idea. It's hard for users to know what to do with it. What if we do this:

1. Add a `hasBeenConsumed` boolean to the `CachedRequestWrapper`
2. Check `size > maxCacheSize && hasBeenConsumed` and then something like:

```
log.error("""
    Attempted to read request body, but body has already been consumed 
    and is not cached. You can increase the request body cache size by calling 
    app.maxBodySizeForRequestCache(newMaxSize)
""")
throw HttpResponseException(413, "Payload Too Large")
```

serii833 commented:
looks very nice
it will stop the execution and will give some info in log

tipsy commented:
Great! Could you update the PR?

serii833 commented:
sure.

tipsy commented:
I still think logging is better than throwing an exception, so I'm going to close this.

serii833 commented:
in my opinion it's better to get an exception than dealing with unexpectable behavior.
but it's up to you to decide.
thanks
